go/cache/cache.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2021 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+// DefaultCacheSize is the default size for a Vitess cache instance.
+// If this value is specified in BYTES, Vitess will use a LFU-based cache that keeps track of the total
+// memory usage of all cached entries accurately. If this value is specified in ENTRIES, Vitess will
+// use the legacy LRU cache implementation which only tracks the amount of entries being stored.
+// Changing this value affects:
+// - the default values for CLI arguments in VTGate
+// - the default values for the config files in VTTablet
+// - the default values for the test caches used in integration and end-to-end tests
+// Regardless of the default value used here, the user can always override Vitess' configuration to
+// force a specific cache type (e.g. when passing a value in ENTRIES to vtgate, the service will use
+// a LRU cache).
+const DefaultCacheSize = SizeInEntries(5000)
+
+// GuessCapacity returns a Capacity value for a cache instance based on the defaults in Vitess
+// and the options passed by the user
+func GuessCapacity(inEntries, inBytes int64) Capacity {
+	switch {
+	// If the default cache has a byte capacity, only override it if the user has explicitly
+	// passed a capacity in entries
+	case DefaultCacheSize.Bytes() != 0:
+		if inEntries != 0 {
+			return SizeInEntries(inEntries)
+		}
+		return SizeInBytes(inBytes)
+
+	// If the default cache has capacity in entries, only override it if the user has explicitly
+	// passed a capacity in bytes
+	case DefaultCacheSize.Entries() != 0:
+		if inBytes != 0 {
+			return SizeInBytes(inBytes)
+		}
+		return SizeInEntries(inEntries)
+
+	default:
+		panic("DefaultCacheSize is not initialized")
+	}
+}
+
+// const DefaultCacheSize = SizeInBytes(64 * 1024 * 1024)
+
+// Cache is a generic interface type for a data structure that keeps recently used
+// objects in memory and evicts them when it becomes full.
+type Cache interface {
+	Get(key string) (interface{}, bool)
+	Set(key string, val interface{}) bool
+	ForEach(callback func(interface{}) bool)
+
+	Delete(key string)
+	Clear()
+	Wait()
```

Comment:
All these methods are pretty self explanatory, and I'm fine with no comments, except for `Wait()` here. This bad boy needs a few words explaining it's purpose.

Reply:
:+1: commented